lib/Protocol/Matrix.pm

```diff
@@ -106,25 +104,52 @@ character string. This is re-exported from L<MIME::Base64> for convenience.
 
    sign_json( $data, secret_key => $key, origin => $name, key_id => $id )
 
+or:
+
+   sign_json( $data,
+      eddsa_secret_key => $secret_key,
+      eddsa_public_key => $public_key,
+      origin => $name,
+      key_id => $id,
+   )
+
 Modifies the given HASH reference in-place to add a signature. This signature
 is created from the given key, and annotated as being from the given origin
 name and key ID. Existing signatures already in the hash are not disturbed.
 
-The C<$key> should be a plain byte string or L<Data::Locker> object obtained
-from L<Crypt::NaCl::Sodium::sign>'s C<keypair> method.
+The key can be specified in one of two ways: either with C<secret_key>, in
+which case the C<$key> should be a plain byte string or L<Data::Locker> object
+obtained from L<Crypt::NaCl::Sodium::sign>'s C<keypair> method; or with
+C<eddsa_secret_key> with a key returned by
+L<Crypt::Ed25519::eddsa_secret_key>. In the latter case C<eddsa_public_key> is
+optional, and the public key will be derived if not given.
 
 =cut
 
 sub sign_json
 {
    my ( $data, %args ) = @_;
 
-   my $key = $args{secret_key} or croak "Require a 'secret_key'";
-
    my $origin = $args{origin} or croak "Require an 'origin'";
    my $key_id = $args{key_id} or croak "Require a 'key_id'";
 
-   my $signature = $sign->mac( encode_json_for_signing( $data ), $key );
+   my ( $secret_key, $public_key );
+   if( exists $args{secret_key} ) {
+      my $key = $args{secret_key};
+
+      # libsodium's "private keys" are actually 64-byte tuples of (seed, public key).
+      # fish them out for use with eddsa_sign.
+      length( $key ) == 64 or croak "secret_key must be 64 bytes";
+      $secret_key = substr( $key, 0, 32 );
+      $public_key = substr( $key, 32, 32 );
```

Comment:
Does this properly handle a `Data::Locker` which the comment above says can be passed in here?

Reply:
bah. No, apparently not. I guess I'll just update the docstring.

Reply:
oh wait, it does, provided the data isn't locked.

```perl
use Crypt::NaCl::Sodium qw(:utils);

my $password = "12345678" x 8;
my $password_locker = Data::BytesLocker->new($password, wipe => 1);

print length($password_locker), "\n";
print substr($password_locker, 0, 32), "\n";
```
->
```
64
12345678123456781234567812345678
```

(I think it means `Data::BytesLocker`, which is what `keypair` actually returns. `Data::Locker` doesn't seem to be a thing)

Reply:
That's surprising, but thanks for checking!